File: quant1x/base/timestamp.py
# -*- coding: UTF-8 -*-
import re
import time
from dataclasses import dataclass
from datetime import datetime
from typing import List

# 仅日期格式: 2022-11-28
FORMAT_ONLY_DATE = '%Y-%m-%d'
# 仅时间格式: 09:15:59
FORMAT_ONLY_TIME = '%H:%M:%S'
# 文件名中包含日期的日期格式: 20221128
FORMAT_FILE_DATE = '%Y%m%d'
# 时间戳: 2022-11-28 09:15:59
FORMAT_DATETIME = '%Y-%m-%d %H:%M:%S'
# 时间戳带毫秒数, 如果毫秒数保留前3位, 需要自己截取: 2022-11-28 09:15:59.123456
FORMAT_TIMESTAMP = '%Y-%m-%d %H:%M:%S.%f'


def seconds_to_timestamp(x: int):
    """
    秒数转时间戳字符串
    :param x:
    :return:
    """
    return time.strftime(FORMAT_DATETIME, time.localtime(x))


@dataclass
class TimeRange(object):
    """
    时间范围, 用~或-间隔HH-MM-SS
    """
    begin: str
    end: str

    def __init__(self, time_range: str):
        """
        构造
        :param time_range:
        :return:
        """
        self.begin = ''
        self.end = ''

        time_range = time_range.strip()
        list = re.split(r"[~-]\s*", time_range)
        if len(list) != 2:
            raise RuntimeError("非法的时间格式")
        # TODO：时间格式校验
        # 时间排序
        self.begin = list[0].strip()
        self.end = list[1].strip()
        if self.begin > self.end:
            self.begin, self.end = self.end, self.begin

    def is_trading(self, timestamp: str = "") -> bool:
        """
        是否交易中
        :param timestamp: %H:%M:%S
        :return:
        """
        timestamp = timestamp.strip()
        if len(timestamp) == 0:
            timestamp = time.strftime(FORMAT_ONLY_TIME)
        if self.begin <= timestamp <= self.end:
            return True
        return False

    def is_valid(self) -> bool:
        """
        时段是否有效
        :return:
        """
        return self.begin != '' and self.end != ''

    def is_session_pre(self, timestamp: str = "") -> bool:
        """
        是否盘前
        :param timestamp: %H:%M:%S
        """
        timestamp = timestamp.strip()
        if len(timestamp) == 0:
            timestamp = time.strftime(FORMAT_ONLY_TIME)
        return timestamp < self.begin

    def is_session_reg(self, timestamp: str = "") -> bool:
        """
        是否盘中
        :param timestamp: %H:%M:%S
        """
        timestamp = timestamp.strip()
        if len(timestamp) == 0:
            timestamp = time.strftime(FORMAT_ONLY_TIME)
        return self.is_trading(timestamp)

    def is_session_post(self, timestamp: str = "") -> bool:
        """
        是否盘后
        :param timestamp: %H:%M:%S
        """
        timestamp = timestamp.strip()
        if len(timestamp) == 0:
            timestamp = time.strftime(FORMAT_ONLY_TIME)
        return timestamp > self.end


@dataclass
class TradingSession:
    """
    交易时段
    """
    sessions: List

    def __init__(self, time_range: str):
        """
        构造
        :param time_range:
        """
        self.sessions = []
        time_range = time_range.strip()
        list = re.split(r",\s*", time_range)
        for v in list:
            v = v.strip()
            r = TimeRange(v)
            self.sessions.append(r)

    def is_trading(self, timestamp: str = "") -> bool:
        """
        是否交易中
        :param timestamp:
        :return:
        """
        timestamp = timestamp.strip()
        if len(timestamp) == 0:
            timestamp = time.strftime(FORMAT_ONLY_TIME)
        for item in self.sessions:
            v = item
            if v.is_trading(timestamp):
                return True
        return False

    def is_valid(self) -> bool:
        """
        时段是否有效
        :return:
        """
        for item in self.sessions:
            if not item.is_valid():
                return False
        return True


if __name__ == '__main__':
    dt = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    dt_ms = datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')
    print(dt)
    print(dt_ms)

    time_range = " 09:30:00 ~ 14:56:30 "
    time_range = " 14:56:30 - 09:30:00 "
    tr = TimeRange(time_range)
    print(tr)

    time_range = "11:30:00 ~ 09:15:00 , 15:00:00 - 13:00:00 "
    time_range = "15:00:00 - 13:00:00 "

    tr = TradingSession(time_range)
    print(tr)
    print(tr.is_trading('09:30:00'))

File: examples/q1x/src/base/timestamp.rs
use chrono::{Local, TimeZone};
use chrono::NaiveTime;
use regex::Regex;
use std::fmt;
use thiserror::Error;

// 时间格式常量
pub const FORMAT_ONLY_DATE: &str = "%Y-%m-%d";
pub const FORMAT_ONLY_TIME: &str = "%H:%M:%S";
#[allow(dead_code)]
pub const FORMAT_FILE_DATE: &str = "%Y%m%d";
pub const FORMAT_DATETIME: &str = "%Y-%m-%d %H:%M:%S";
#[allow(dead_code)]
pub const FORMAT_TIMESTAMP: &str = "%Y-%m-%d %H:%M:%S.%f";

/// 时间转换错误类型
#[derive(Debug, Error)]
pub enum TimeError {
    #[error("Invalid timestamp format")]
    InvalidFormat,
    #[error("Parse error: {0}")]
    ParseError(#[from] chrono::ParseError),
}

/// 时间范围错误类型
#[derive(Debug, Error)]
pub enum TimeRangeError {
    #[error("Invalid time range format")]
    InvalidFormat,
    #[error("Parse error: {0}")]
    TimeError(#[from] TimeError),
}

/// 交易时段结构体
#[derive(Debug, PartialEq)]
pub struct TimeRange {
    begin: NaiveTime,
    end: NaiveTime,
}

impl TimeRange {
    pub fn new(time_range: &str) -> Result<Self, TimeRangeError> {
        let re = Regex::new(r"[~-]\s*").unwrap();
        let parts: Vec<&str> = re.split(time_range.trim()).collect();
        if parts.len() != 2 {
            return Err(TimeRangeError::InvalidFormat);
        }

        let parse_time = |s: &str| -> Result<NaiveTime, TimeError> {
            NaiveTime::parse_from_str(s.trim(), FORMAT_ONLY_TIME)
                .map_err(|_| TimeError::InvalidFormat)
        };

        let mut begin = parse_time(parts[0])?;
        let mut end = parse_time(parts[1])?;

        if begin > end {
            std::mem::swap(&mut begin, &mut end);
        }

        Ok(Self { begin, end })
    }

    /// 是否在交易中
    pub fn is_trading(&self, timestamp: Option<&str>) -> Result<bool, TimeError> {
        let time_str = match timestamp {
            Some(ts) => ts.trim(),
            None => &NaiveTime::now().format(FORMAT_ONLY_TIME).to_string(),
        };

        let time = NaiveTime::parse_from_str(time_str, FORMAT_ONLY_TIME)?;
        Ok(time >= self.begin && time <= self.end)
    }

    pub fn is_valid(&self) -> bool {
        !self.begin.format(FORMAT_ONLY_TIME).to_string().is_empty() &&
            !self.end.format(FORMAT_ONLY_TIME).to_string().is_empty()
    }

    /// 盘前
    pub fn is_session_pre(&self, timestamp: Option<&str>) -> Result<bool, TimeError> {
        let time_str = match timestamp {
            Some(ts) => ts.trim(),
            None => &NaiveTime::now().format(FORMAT_ONLY_TIME).to_string(),
        };

        let time = NaiveTime::parse_from_str(time_str, FORMAT_ONLY_TIME)?;
        Ok(time < self.begin)
    }

    /// 盘后
    pub fn is_session_post(&self, timestamp: Option<&str>) -> Result<bool, TimeError> {
        let time_str = match timestamp {
            Some(ts) => ts.trim(),
            None => &NaiveTime::now().format(FORMAT_ONLY_TIME).to_string(),
        };

        let time = NaiveTime::parse_from_str(time_str, FORMAT_ONLY_TIME)?;
        Ok(time > self.end)
    }
}

impl fmt::Display for TimeRange {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ~ {}", self.begin.format(FORMAT_ONLY_TIME), self.end.format(FORMAT_ONLY_TIME))
    }
}

/// 交易时段集合
#[derive(Debug)]
pub struct TradingSession {
    sessions: Vec<TimeRange>,
}

impl TradingSession {
    pub fn new(time_ranges: &str) -> Result<Self, TimeRangeError> {
        let re = Regex::new(r",\s*").unwrap();
        let parts: Vec<&str> = re.split(time_ranges.trim()).collect();

        let mut sessions = Vec::new();
        for part in parts {
            let tr = TimeRange::new(part)?;
            sessions.push(tr);
        }
        Ok(Self { sessions })
    }

    pub fn is_trading(&self, timestamp: Option<&str>) -> Result<bool, TimeError> {
        for session in &self.sessions {
            if session.is_trading(timestamp)? {
                return Ok(true);
            }
        }
        Ok(false)
    }

    pub fn is_valid(&self) -> bool {
        self.sessions.iter().all(|tr| tr.is_valid())
    }
}

impl fmt::Display for TradingSession {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "[{}]",
            self.sessions
                .iter()
                .map(|tr| tr.to_string())
                .collect::<Vec<String>>()
                .join(", ")
        )
    }
}

/// 时间转换函数
// pub fn seconds_to_timestamp(seconds: i64) -> String {
//     NaiveDateTime::from_timestamp_opt(seconds, 0)
//         .map(|dt| dt.format(FORMAT_DATETIME).to_string())
//         .unwrap_or_else(|| {
//             eprintln!("Invalid timestamp: {}", seconds);
//             "Invalid timestamp".to_string()
//         })
// }

// 安全且无警告的时间转换方法
pub fn seconds_to_timestamp(seconds: i64) -> String {
    // 创建本地时间对象
    let dt = Local.timestamp_opt(seconds, 0)
        .single()
        .expect("无效时间戳");
    //println!("{}", dt);
    // 提取无时区信息的时间
    let dt1 = dt.naive_local();
    //println!("{}", dt1);
    //println!("{}", Local::now().format(FORMAT_DATETIME).to_string());
    //println!("{}", dt.offset());
    dt1.format(FORMAT_DATETIME).to_string()
}


use std::time::{SystemTime, UNIX_EPOCH};

// 直接本地时间转换(无需UTC)
pub fn local_timestamp_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("Time went backwards")
        .as_millis() as i64
}

// // 将秒数转换为本地时间
// pub fn seconds_to_local_datetime(seconds: i64) -> NaiveDateTime {
//     let d = std::time::Duration::from_secs(seconds as u64);
//     let systime = UNIX_EPOCH + d;
//     systime
//         .duration_since(UNIX_EPOCH)
//         .map(|dur| NaiveDateTime::from_timestamp_opt(dur.as_secs() as i64, 0))
//         .unwrap()
//         .expect("Invalid timestamp")
// }


/// 扩展NaiveTime的now()方法
trait NaiveTimeExt {
    fn now() -> NaiveTime;
}

impl NaiveTimeExt for NaiveTime {
    fn now() -> Self {
        chrono::Local::now().naive_local().time()
    }
}

/// 单元测试模块
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_time_conversion() {
        let now = chrono::Local::now().timestamp();
        let timestamp = seconds_to_timestamp(now);
        assert!(timestamp.len() >= 19);
        let timestamp = 1697004000; // 北京时间2023-10-10 12:00:00
        //println!("Timestamp: {}", timestamp_to_local(timestamp));
        println!("Timestamp: {}", seconds_to_timestamp(timestamp));
    }

    #[test]
    fn test_time_range_parsing() {
        let tr = TimeRange::new("09:30:00 ~ 14:56:30").unwrap();
        assert_eq!(tr.begin.format(FORMAT_ONLY_TIME).to_string(), "09:30:00");
        assert_eq!(tr.end.format(FORMAT_ONLY_TIME).to_string(), "14:56:30");

        let tr = TimeRange::new("14:56:30 - 09:30:00").unwrap();
        assert_eq!(tr.begin.format(FORMAT_ONLY_TIME).to_string(), "09:30:00");
        assert_eq!(tr.end.format(FORMAT_ONLY_TIME).to_string(), "14:56:30");
    }

    #[test]
    fn test_trading_status() {
        let tr = TimeRange::new("09:30:00 ~ 14:56:30").unwrap();
        //assert!(tr.is_trading().unwrap());
        assert!(tr.is_trading(Some("12:00:00")).unwrap());
        assert!(!tr.is_trading(Some("08:00:00")).unwrap());
        assert!(!tr.is_trading(Some("15:00:00")).unwrap());
    }

    #[test]
    fn test_trading_session() {
        let ts = TradingSession::new("11:30:00 ~ 09:15:00, 15:00:00 - 13:00:00").unwrap();
        assert!(ts.is_trading(Some("10:00:00")).unwrap());
        assert!(ts.is_trading(Some("14:00:00")).unwrap());
        assert!(!ts.is_trading(Some("12:00:00")).unwrap());
    }

    #[test]
    fn test_edge_cases() {
        let tr = TimeRange::new("23:00:00 ~ 01:00:00").unwrap();
        assert!(!tr.is_trading(Some("23:30:00")).unwrap());
        assert!(!tr.is_trading(Some("00:30:00")).unwrap());
    }
}